File: Windows/src/Widget.TimeTracking.Infrastructure.Mock/Services/LocalJsonTimeTrackingService.cs
using Widget.TimeTracking.Core.Enums;
using Widget.TimeTracking.Core.Models;
using Widget.TimeTracking.Core.Results;
using Widget.TimeTracking.Core.Rules;
using Widget.TimeTracking.Core.Services;
using Widget.TimeTracking.Infrastructure.Mock.Configuration;
using Widget.TimeTracking.Infrastructure.Mock.Persistence;

namespace Widget.TimeTracking.Infrastructure.Mock.Services;

public sealed class LocalJsonTimeTrackingService : ITimeTrackingService
{
    private readonly JsonTimeTrackingStore _store;
    private readonly LocalStateOptions _options;

    public LocalJsonTimeTrackingService(JsonTimeTrackingStore store, LocalStateOptions options)
    {
        _store = store;
        _options = options;
    }

    public async Task<TimeTrackingSnapshot> GetStateAsync(CancellationToken cancellationToken = default)
    {
        var document = await _store.LoadAsync(cancellationToken);
        return ToSnapshot(document, DateTimeOffset.UtcNow);
    }

    public Task<TimeTrackingCommandResult> ClockInAsync(CancellationToken cancellationToken = default) =>
        ExecuteAsync(
            TimeTrackingAction.ClockIn,
            WorkdayEventType.ClockIn,
            cancellationToken,
            onSuccess: (document, nowUtc) =>
            {
                document.CurrentShiftStartedAtUtc = nowUtc;
                document.ActiveBreakType = BreakType.None;
            });

    public Task<TimeTrackingCommandResult> StartBreakAsync(CancellationToken cancellationToken = default) =>
        StartCoffeeBreakAsync(cancellationToken);

    public async Task<TimeTrackingCommandResult> EndBreakAsync(CancellationToken cancellationToken = default)
    {
        var document = await _store.LoadAsync(cancellationToken);

        return document.ActiveBreakType switch
        {
            BreakType.Coffee => await EndCoffeeBreakAsync(cancellationToken),
            BreakType.Food => await EndFoodBreakAsync(cancellationToken),
            _ => await BuildFailureAsync(document, "No hay un descanso activo para finalizar.", cancellationToken)
        };
    }

    public Task<TimeTrackingCommandResult> StartCoffeeBreakAsync(CancellationToken cancellationToken = default) =>
        ExecuteAsync(
            TimeTrackingAction.StartCoffeeBreak,
            WorkdayEventType.StartCoffeeBreak,
            cancellationToken,
            onSuccess: (document, nowUtc) =>
            {
                document.ActiveBreakType = BreakType.Coffee;
                document.BreakSessions.Add(new BreakSession(BreakType.Coffee, nowUtc, null));
            });

    public Task<TimeTrackingCommandResult> EndCoffeeBreakAsync(CancellationToken cancellationToken = default) =>
        ExecuteAsync(
            TimeTrackingAction.EndCoffeeBreak,
            WorkdayEventType.EndCoffeeBreak,
            cancellationToken,
            onSuccess: (document, nowUtc) =>
            {
                CloseActiveBreak(document, BreakType.Coffee, nowUtc);
                document.ActiveBreakType = BreakType.None;
            });

    public Task<TimeTrackingCommandResult> StartFoodBreakAsync(CancellationToken cancellationToken = default) =>
        ExecuteAsync(
            TimeTrackingAction.StartFoodBreak,
            WorkdayEventType.StartFoodBreak,
            cancellationToken,
            onSuccess: (document, nowUtc) =>
            {
                document.ActiveBreakType = BreakType.Food;
                document.BreakSessions.Add(new BreakSession(BreakType.Food, nowUtc, null));
            });

    public Task<TimeTrackingCommandResult> EndFoodBreakAsync(CancellationToken cancellationToken = default) =>
        ExecuteAsync(
            TimeTrackingAction.EndFoodBreak,
            WorkdayEventType.EndFoodBreak,
            cancellationToken,
            onSuccess: (document, nowUtc) =>
            {
                CloseActiveBreak(document, BreakType.Food, nowUtc);
                document.ActiveBreakType = BreakType.None;
            });

    public Task<TimeTrackingCommandResult> ClockOutAsync(CancellationToken cancellationToken = default) =>
        ExecuteAsync(
            TimeTrackingAction.ClockOut,
            WorkdayEventType.ClockOut,
            cancellationToken,
            onSuccess: (document, nowUtc) =>
            {
                CloseActiveBreakIfAny(document, nowUtc);

                if (document.CurrentShiftStartedAtUtc.HasValue)
                {
                    var completedDuration = CalculateWorkedDurationForShift(document, document.CurrentShiftStartedAtUtc.Value, nowUtc);
                    document.LastCompletedShiftWorkedSeconds = (long)completedDuration.TotalSeconds;
                    document.WorkedThisMonthSeconds += (long)completedDuration.TotalSeconds;
                }

                document.CurrentShiftStartedAtUtc = null;
                document.ActiveBreakType = BreakType.None;
            });

    private async Task<TimeTrackingCommandResult> ExecuteAsync(
        TimeTrackingAction action,
        WorkdayEventType eventType,
        CancellationToken cancellationToken,
        Action<TimeTrackingStateDocument, DateTimeOffset> onSuccess)
    {
        var document = await _store.LoadAsync(cancellationToken);
        var nowUtc = DateTimeOffset.UtcNow;

        if (!TimeTrackingStateMachine.TryTransition(
                document.Status,
                document.ActiveBreakType,
                action,
                out var nextStatus,
                out var errorMessage))
        {
            return await BuildFailureAsync(document, errorMessage ?? "La acción no es válida.", cancellationToken);
        }

        var breakTypeForAction = ResolveBreakTypeForAction(action, document.ActiveBreakType);
        onSuccess(document, nowUtc);

        document.Status = nextStatus;
        document.LastAction = action;
        document.LastActionAtUtc = nowUtc;
        document.History.Add(new TimeTrackingHistoryEntry(nowUtc, action, nextStatus, breakTypeForAction));
        document.WorkdayEvents.Add(new WorkdayEvent(nowUtc, eventType, nextStatus, breakTypeForAction));

        TrimHistory(document);
        await _store.SaveAsync(document, cancellationToken);

        return TimeTrackingCommandResult.Success(ToSnapshot(document, nowUtc));
    }

    private async Task<TimeTrackingCommandResult> BuildFailureAsync(
        TimeTrackingStateDocument document,
        string message,
        CancellationToken cancellationToken)
    {
        var snapshot = ToSnapshot(document, DateTimeOffset.UtcNow);
        await Task.CompletedTask;
        return TimeTrackingCommandResult.Failure(snapshot, message);
    }

    private TimeTrackingSnapshot ToSnapshot(TimeTrackingStateDocument document, DateTimeOffset nowUtc)
    {
        var currentDay = DateOnly.FromDateTime(nowUtc.LocalDateTime);
        var activeBreak = document.BreakSessions.LastOrDefault(session => session.IsActive);
        var todayBreakSessions = document.BreakSessions
            .Where(session => DateOnly.FromDateTime(session.StartedAtUtc.LocalDateTime) == currentDay)
            .OrderBy(session => session.StartedAtUtc)
            .ToArray();
        var todayEvents = document.WorkdayEvents
            .Where(item => DateOnly.FromDateTime(item.OccurredAtUtc.LocalDateTime) == currentDay)
            .OrderBy(item => item.OccurredAtUtc)
            .ToArray();
        var currentShiftWorked = document.CurrentShiftStartedAtUtc.HasValue
            ? CalculateWorkedDurationForShift(document, document.CurrentShiftStartedAtUtc.Value, nowUtc)
            : TimeSpan.Zero;
        var summary = new DailyWorkSummary(
            Day: currentDay,
            CurrentShiftWorkedDuration: currentShiftWorked,
            LastCompletedShiftWorkedDuration: TimeSpan.FromSeconds(document.LastCompletedShiftWorkedSeconds),
            WorkedThisMonthDuration: TimeSpan.FromSeconds(document.WorkedThisMonthSeconds) + currentShiftWorked,
            CoffeeBreakDurationToday: SumBreakDuration(todayBreakSessions, BreakType.Coffee, nowUtc),
            FoodBreakDurationToday: SumBreakDuration(todayBreakSessions, BreakType.Food, nowUtc));

        return new TimeTrackingSnapshot(
            Status: document.Status,
            LastAction: document.LastAction,
            LastActionAtUtc: document.LastActionAtUtc,
            ActiveBreakType: document.ActiveBreakType,
            ActiveBreak: activeBreak,
            History: document.History
                .OrderByDescending(item => item.OccurredAtUtc)
                .ToArray(),
            WorkdayEvents: todayEvents,
            BreakSessions: todayBreakSessions,
            Summary: summary,
            AvailableActions: TimeTrackingStateMachine.GetAvailableActions(document.Status, document.ActiveBreakType));
    }

    private static TimeSpan SumBreakDuration(IEnumerable<BreakSession> sessions, BreakType type, DateTimeOffset nowUtc) =>
        sessions
            .Where(session => session.Type == type)
            .Aggregate(TimeSpan.Zero, (total, session) => total + session.GetDuration(nowUtc));

    private static void CloseActiveBreak(TimeTrackingStateDocument document, BreakType breakType, DateTimeOffset nowUtc)
    {
        var activeBreakIndex = document.BreakSessions.FindLastIndex(session => session.IsActive && session.Type == breakType);
        if (activeBreakIndex < 0)
        {
            return;
        }

        var activeBreak = document.BreakSessions[activeBreakIndex];
        document.BreakSessions[activeBreakIndex] = activeBreak with { EndedAtUtc = nowUtc };
    }

    private static void CloseActiveBreakIfAny(TimeTrackingStateDocument document, DateTimeOffset nowUtc)
    {
        var activeBreakIndex = document.BreakSessions.FindLastIndex(session => session.IsActive);
        if (activeBreakIndex < 0)
        {
            return;
        }

        var activeBreak = document.BreakSessions[activeBreakIndex];
        document.BreakSessions[activeBreakIndex] = activeBreak with { EndedAtUtc = nowUtc };
    }

    private static TimeSpan CalculateWorkedDurationForShift(
        TimeTrackingStateDocument document,
        DateTimeOffset shiftStartUtc,
        DateTimeOffset nowUtc)
    {
        if (nowUtc <= shiftStartUtc)
        {
            return TimeSpan.Zero;
        }

        var grossDuration = nowUtc - shiftStartUtc;
        var breakDuration = document.BreakSessions
            .Where(session => session.StartedAtUtc >= shiftStartUtc)
            .Aggregate(TimeSpan.Zero, (total, session) => total + session.GetDuration(nowUtc));

        var netDuration = grossDuration - breakDuration;
        return netDuration < TimeSpan.Zero ? TimeSpan.Zero : netDuration;
    }

    private void TrimHistory(TimeTrackingStateDocument document)
    {
        if (document.History.Count > _options.MaxHistoryEntries)
        {
            document.History = document.History
                .OrderByDescending(item => item.OccurredAtUtc)
                .Take(_options.MaxHistoryEntries)
                .OrderBy(item => item.OccurredAtUtc)
                .ToList();
        }

        var cutoffDate = DateTimeOffset.UtcNow.AddDays(-30);
        document.WorkdayEvents = document.WorkdayEvents
            .Where(item => item.OccurredAtUtc >= cutoffDate)
            .OrderBy(item => item.OccurredAtUtc)
            .ToList();
        document.BreakSessions = document.BreakSessions
            .Where(item => item.StartedAtUtc >= cutoffDate)
            .OrderBy(item => item.StartedAtUtc)
            .ToList();
    }

    private static BreakType ResolveBreakTypeForAction(TimeTrackingAction action, BreakType activeBreakTypeBeforeAction) =>
        action switch
        {
            TimeTrackingAction.StartCoffeeBreak or TimeTrackingAction.EndCoffeeBreak => BreakType.Coffee,
            TimeTrackingAction.StartFoodBreak or TimeTrackingAction.EndFoodBreak => BreakType.Food,
            TimeTrackingAction.StartBreak or TimeTrackingAction.EndBreak => activeBreakTypeBeforeAction,
            _ => BreakType.None
        };
}

File: Mac/widget/widget/Services/LocalJsonTimeTrackingService.swift
import Foundation

final class LocalJsonTimeTrackingService: @unchecked Sendable {
    private let store = JsonTimeTrackingStore()

    func getState() throws -> TimeTrackingSnapshot {
        let document = try store.load()
        return toSnapshot(document: document, nowUtc: Date())
    }

    func clockIn() throws -> TimeTrackingCommandResult {
        try execute(
            action: .clockIn,
            eventType: .clockIn,
            onSuccess: { doc, nowUtc in
                doc.currentShiftStartedAtUtc = nowUtc
                doc.activeBreakType = .none
            }
        )
    }

    func startBreak() throws -> TimeTrackingCommandResult {
        try startCoffeeBreak()
    }

    func endBreak() throws -> TimeTrackingCommandResult {
        var document = try store.load()
        switch document.activeBreakType {
        case .coffee:
            return try endCoffeeBreak()
        case .food:
            return try endFoodBreak()
        case .none:
            return try buildFailure(document, "No hay un descanso activo para finalizar.")
        }
    }

    func startCoffeeBreak() throws -> TimeTrackingCommandResult {
        try execute(
            action: .startCoffeeBreak,
            eventType: .startCoffeeBreak,
            onSuccess: { doc, nowUtc in
                doc.activeBreakType = .coffee
                doc.breakSessions.append(BreakSession(type: .coffee, startedAtUtc: nowUtc, endedAtUtc: nil))
            }
        )
    }

    func endCoffeeBreak() throws -> TimeTrackingCommandResult {
        try execute(
            action: .endCoffeeBreak,
            eventType: .endCoffeeBreak,
            onSuccess: { doc, nowUtc in
                closeActiveBreak(document: &doc, breakType: .coffee, nowUtc: nowUtc)
                doc.activeBreakType = .none
            }
        )
    }

    func startFoodBreak() throws -> TimeTrackingCommandResult {
        try execute(
            action: .startFoodBreak,
            eventType: .startFoodBreak,
            onSuccess: { doc, nowUtc in
                doc.activeBreakType = .food
                doc.breakSessions.append(BreakSession(type: .food, startedAtUtc: nowUtc, endedAtUtc: nil))
            }
        )
    }

    func endFoodBreak() throws -> TimeTrackingCommandResult {
        try execute(
            action: .endFoodBreak,
            eventType: .endFoodBreak,
            onSuccess: { doc, nowUtc in
                closeActiveBreak(document: &doc, breakType: .food, nowUtc: nowUtc)
                doc.activeBreakType = .none
            }
        )
    }

    func clockOut() throws -> TimeTrackingCommandResult {
        try execute(
            action: .clockOut,
            eventType: .clockOut,
            onSuccess: { doc, nowUtc in
                closeActiveBreakIfAny(document: &doc, nowUtc: nowUtc)
                if let start = doc.currentShiftStartedAtUtc {
                    let completed = calculateWorkedDurationForShift(document: doc, shiftStartUtc: start, nowUtc: nowUtc)
                    doc.lastCompletedShiftWorkedSeconds = Int64(completed)
                    doc.workedThisMonthSeconds += Int64(completed)
                }
                doc.currentShiftStartedAtUtc = nil
                doc.activeBreakType = .none
            }
        )
    }

    private func execute(
        action: TimeTrackingAction,
        eventType: WorkdayEventType,
        onSuccess: (inout TimeTrackingStateDocument, Date) -> Void
    ) throws -> TimeTrackingCommandResult {
        var document = try store.load()
        let nowUtc = Date()

        switch TimeTrackingStateMachine.tryTransition(
            currentStatus: document.status,
            activeBreakType: document.activeBreakType,
            action: action
        ) {
        case .failure(let err):
            return try buildFailure(document, err.message)
        case .success(let nextStatus):
            let breakBefore = document.activeBreakType
            onSuccess(&document, nowUtc)
            let breakForAction = resolveBreakTypeForAction(action: action, activeBreakTypeBeforeAction: breakBefore)
            document.status = nextStatus
            document.lastAction = action
            document.lastActionAtUtc = nowUtc
            document.history.append(
                TimeTrackingHistoryEntry(
                    occurredAtUtc: nowUtc,
                    action: action,
                    status: nextStatus,
                    breakType: breakForAction
                )
            )
            document.workdayEvents.append(
                WorkdayEvent(
                    occurredAtUtc: nowUtc,
                    eventType: eventType,
                    statusAfterEvent: nextStatus,
                    breakType: breakForAction
                )
            )
            trimHistory(&document)
            try store.save(document)
            return .success(toSnapshot(document: document, nowUtc: nowUtc))
        }
    }

    private func buildFailure(_ document: TimeTrackingStateDocument, _ message: String) throws -> TimeTrackingCommandResult {
        .failure(toSnapshot(document: document, nowUtc: Date()), message)
    }

    private func toSnapshot(document: TimeTrackingStateDocument, nowUtc: Date) -> TimeTrackingSnapshot {
        let calendar = Calendar.current
        let dayStart = calendar.startOfDay(for: nowUtc)
        let activeBreak = document.breakSessions.last { $0.isActive }
        let breakSessionsForSummary = getBreakSessionsForSummary(document: document, nowUtc: nowUtc)
        let todayEvents = document.workdayEvents
            .filter { calendar.isDate($0.occurredAtUtc, inSameDayAs: nowUtc) }
            .sorted { $0.occurredAtUtc < $1.occurredAtUtc }

        let currentShiftWorked: TimeInterval
        if let start = document.currentShiftStartedAtUtc {
            currentShiftWorked = calculateWorkedDurationForShift(document: document, shiftStartUtc: start, nowUtc: nowUtc)
        } else {
            currentShiftWorked = 0
        }

        let summary = DailyWorkSummary(
            dayStart: dayStart,
            currentShiftWorkedDuration: currentShiftWorked,
            lastCompletedShiftWorkedDuration: TimeInterval(document.lastCompletedShiftWorkedSeconds),
            workedThisMonthDuration: TimeInterval(document.workedThisMonthSeconds) + currentShiftWorked,
            coffeeBreakDurationToday: sumBreakDuration(sessions: breakSessionsForSummary, type: .coffee, nowUtc: nowUtc),
            foodBreakDurationToday: sumBreakDuration(sessions: breakSessionsForSummary, type: .food, nowUtc: nowUtc)
        )

        return TimeTrackingSnapshot(
            status: document.status,
            lastAction: document.lastAction,
            lastActionAtUtc: document.lastActionAtUtc,
            currentShiftStartedAtUtc: document.currentShiftStartedAtUtc,
            activeBreakType: document.activeBreakType,
            activeBreak: activeBreak,
            history: document.history.sorted { $0.occurredAtUtc > $1.occurredAtUtc },
            workdayEvents: todayEvents,
            breakSessions: breakSessionsForSummary,
            summary: summary,
            availableActions: TimeTrackingStateMachine.availableActions(
                status: document.status,
                activeBreakType: document.activeBreakType
            )
        )
    }

    private func getBreakSessionsForSummary(document: TimeTrackingStateDocument, nowUtc: Date) -> [BreakSession] {
        let calendar = Calendar.current
        if let shiftStart = document.currentShiftStartedAtUtc {
            return document.breakSessions
                .filter { $0.startedAtUtc >= shiftStart }
                .sorted { $0.startedAtUtc < $1.startedAtUtc }
        }

        let clockOutsToday = document.workdayEvents
            .filter { $0.eventType == .clockOut && calendar.isDate($0.occurredAtUtc, inSameDayAs: nowUtc) }
            .sorted { $0.occurredAtUtc < $1.occurredAtUtc }

        if let lastClockOut = clockOutsToday.last {
            let clockInsToday = document.workdayEvents
                .filter {
                    $0.eventType == .clockIn
                        && calendar.isDate($0.occurredAtUtc, inSameDayAs: nowUtc)
                        && $0.occurredAtUtc <= lastClockOut.occurredAtUtc
                }
                .sorted { $0.occurredAtUtc < $1.occurredAtUtc }

            if let lastClockIn = clockInsToday.last {
                let shiftStartUtc = lastClockIn.occurredAtUtc
                let shiftEndUtc = lastClockOut.occurredAtUtc
                return document.breakSessions
                    .filter { $0.startedAtUtc >= shiftStartUtc && $0.startedAtUtc <= shiftEndUtc }
                    .sorted { $0.startedAtUtc < $1.startedAtUtc }
            }
        }

        return document.breakSessions
            .filter { calendar.isDate($0.startedAtUtc, inSameDayAs: nowUtc) }
            .sorted { $0.startedAtUtc < $1.startedAtUtc }
    }

    private func sumBreakDuration(sessions: [BreakSession], type: BreakType, nowUtc: Date) -> TimeInterval {
        sessions.filter { $0.type == type }.reduce(0) { $0 + $1.duration(nowUtc: nowUtc) }
    }

    private func closeActiveBreak(document: inout TimeTrackingStateDocument, breakType: BreakType, nowUtc: Date) {
        guard let index = document.breakSessions.lastIndex(where: { $0.isActive && $0.type == breakType }) else { return }
        var session = document.breakSessions[index]
        session.endedAtUtc = nowUtc
        document.breakSessions[index] = session
    }

    private func closeActiveBreakIfAny(document: inout TimeTrackingStateDocument, nowUtc: Date) {
        guard let index = document.breakSessions.lastIndex(where: { $0.isActive }) else { return }
        var session = document.breakSessions[index]
        session.endedAtUtc = nowUtc
        document.breakSessions[index] = session
    }

    private func calculateWorkedDurationForShift(
        document: TimeTrackingStateDocument,
        shiftStartUtc: Date,
        nowUtc: Date
    ) -> TimeInterval {
        TimeTrackingLiveDuration.netWorkedSeconds(
            shiftStartUtc: shiftStartUtc,
            breakSessions: document.breakSessions,
            nowUtc: nowUtc
        )
    }

    private func trimHistory(_ document: inout TimeTrackingStateDocument) {
        if document.history.count > LocalStatePaths.maxHistoryEntries {
            document.history = document.history
                .sorted { $0.occurredAtUtc > $1.occurredAtUtc }
                .prefix(LocalStatePaths.maxHistoryEntries)
                .sorted { $0.occurredAtUtc < $1.occurredAtUtc }
        }

        let cutoff = Calendar.current.date(byAdding: .day, value: -30, to: Date()) ?? Date.distantPast
        document.workdayEvents = document.workdayEvents
            .filter { $0.occurredAtUtc >= cutoff }
            .sorted { $0.occurredAtUtc < $1.occurredAtUtc }
        document.breakSessions = document.breakSessions
            .filter { $0.startedAtUtc >= cutoff }
            .sorted { $0.startedAtUtc < $1.startedAtUtc }
    }

    private func resolveBreakTypeForAction(action: TimeTrackingAction, activeBreakTypeBeforeAction: BreakType) -> BreakType {
        switch action {
        case .startCoffeeBreak, .endCoffeeBreak:
            return .coffee
        case .startFoodBreak, .endFoodBreak:
            return .food
        case .startBreak, .endBreak:
            return activeBreakTypeBeforeAction
        default:
            return .none
        }
    }
}
